gh-122288: Improve performances of `fnmatch.translate`

picnixz opened the pull request:
This is a smaller PR compared to the one for the C implementation and is probably easier to review. Below are the benchmarks reported on the issue:

```text
+------------------------------------------+-----------------------+-----------------------+
| Benchmark                                | fnmatch-translate-ref | fnmatch-translate-py  |
+==========================================+=======================+=======================+
| abc/[!]b-ac-z9-1]/def/\*?/*/**c/?*[][!]  | 6.09 us               | 3.99 us: 1.53x faster |
+------------------------------------------+-----------------------+-----------------------+
| !abc/[!]b-ac-z9-1]/def/\*?/*/**c/?*[][!] | 6.39 us               | 4.07 us: 1.57x faster |
+------------------------------------------+-----------------------+-----------------------+
| a**?**cd**?**??k***                      | 2.24 us               | 1.51 us: 1.49x faster |
+------------------------------------------+-----------------------+-----------------------+
| a/**/b/**/c                              | 1.97 us               | 1.12 us: 1.76x faster |
+------------------------------------------+-----------------------+-----------------------+
| man/man1/bash.1                          | 3.00 us               | 1.21 us: 2.48x faster |
+------------------------------------------+-----------------------+-----------------------+
| a*b*c*d*e*f*g*h*i*j*k*l*m*n*o**          | 5.40 us               | 3.33 us: 1.62x faster |
+------------------------------------------+-----------------------+-----------------------+
| Geometric mean                           | (ref)                 | 1.71x faster          |
+------------------------------------------+-----------------------+-----------------------+
```

<!-- gh-issue-number: gh-122288 -->
* Issue: gh-122288
<!-- /gh-issue-number -->

picnixz committed:
improve performances of `fnmatch.translate`

picnixz committed:
add tests

picnixz committed:
blurb

picnixz committed:
fix usages

picnixz committed:
keep legacy version for glob

picnixz committed:
actually not needed...

picnixz committed:
Merge branch 'main' into fnmatch-py-implementation

picnixz committed:
reduce the number of calls to `str.join`

picnixz commented on Lib/fnmatch.py:
```diff
@@ -143,50 +150,33 @@ def _translate(pat, STAR, QUESTION_MARK):
                     add('.')
                 else:
+                    # Escape set operations (&&, ~~ and ||).
+                    stuff = _set_ops_re.sub(r'\\\1', stuff)
```
I could also cache the call to `sub` as I said to Serhiy in another PR. Will do it tomorrow.

picnixz committed:
micro-optimization on `re.sub`

barneygale commented on Lib/fnmatch.py:
```diff
@@ -143,50 +150,33 @@ def _translate(pat, STAR, QUESTION_MARK):
+        return fr'(?s:{"".join(parts)})\Z'
+    iter_indices = iter(indices)
+    i, j = 0, next(iter_indices)
+    buffer = parts[i:j]
```
```suggestion
    j = next(iter_indices)
    buffer = parts[:j]
```

barneygale commented on Lib/fnmatch.py:
```diff
@@ -143,50 +150,33 @@ def _translate(pat, STAR, QUESTION_MARK):
-    while i < n and inp[i] is not STAR:
-        add(inp[i])
-        i += 1
-    # Now deal with STAR fixed STAR fixed ...
-    # For an interior `STAR fixed` pairing, we want to do a minimal
-    # .*? match followed by `fixed`, with no possibility of backtracking.
-    # Atomic groups ("(?>...)") allow us to spell that directly.
-    # Note: people rely on the undocumented ability to join multiple
-    # translate() results together via "|" to build large regexps matching
-    # "one of many" shell patterns.
```
It would be good to preserve this helpful comment.

barneygale commented on Lib/fnmatch.py:
```diff
@@ -77,23 +77,32 @@ def translate(pat):
+_re_setops_sub = re.compile(r'([&~|])').sub
+_re_escape = functools.lru_cache(maxsize=32768)(re.escape)
 
 def _translate(pat, STAR, QUESTION_MARK):
```
```suggestion
def _translate(pat, star, question_mark):
```

Now that `STAR` is no longer `object()`, this naming is vestigial.

picnixz replied:
The thing is, it's still possible for other parts of the library to use different values for `STAR`. I'd like to keep it that way (e.g., `results.extend(fnmatch._translate(part, f'{not_sep}*', not_sep)[0])` in glob.py, the star is not just a "star", it's a bit more than that I'd say).

dg-pb replied:
I think this was more about making argument names lower case than anything else.

picnixz replied:
Mmh. I like the fact that it's all caps to highlight the fact that it's a constant, as if `_translate` was a macro in this case. But maybe it's better to drop this historical approach and move to the modern one.

barneygale replied:
The thing is, `STAR` _used_ to be an `object()` constant in `fnmatch`, but I changed that when I introduced `glob.translate()`. I should have changed them to lower case in that PR, and I regret not doing so now.

picnixz replied:
I see. I hope the last commit solved your regrets :)

barneygale commented on Lib/fnmatch.py:
```diff
@@ -77,23 +77,32 @@ def translate(pat):
         if c == '*':
+            # store the position of the wildcard
+            indices.append(len(res))
+            add(STAR)
```
```suggestion
            add(star)
```

barneygale commented on Lib/fnmatch.py:
```diff
@@ -77,23 +77,32 @@ def translate(pat):
+            # Handling '?' one at a time seems to more efficient
+            # even if there are consecutive '?' that could have
+            # been written directly.
             add(QUESTION_MARK)
```
```suggestion
            add(question_mark)
```

barneygale commented on Lib/test/test_fnmatch.py:
```diff
@@ -250,6 +250,72 @@ def test_translate(self):
+            ('ab*cd*12*34*', r'(?s:ab(?>.*?cd)(?>.*?12)(?>.*?34).*)\Z'),
+        ]:
+            translated = translate(pattern)
+            self.assertEqual(translated, expect, pattern)
```
Using a `for` loop like this makes the tracebacks slightly less useful when a test case fails. Either use a [subtest](https://docs.python.org/3/library/unittest.html#subtests) or spell each assertion out explicitly (e.g. `self.assertEqual(translate('ab*'), r'(?s:ab.*)\Z')`).

(Also below)

picnixz replied:
I forgot to put the `subTest`. Thanks for reminding me of that!

barneygale reviewed (commented):
This is a nice improvement :)

picnixz committed:
address Barney's review

dg-pb commented on Lib/fnmatch.py:
```diff
@@ -77,23 +77,32 @@ def translate(pat):
         i = i+1
         if c == '*':
+            # store the position of the wildcard
+            indices.append(len(res))
```
Can factor out `getattr`

dg-pb replied:
Also `chunks.append` below.

picnixz replied:
I doubt it will change much and I wouldn't need to add an aditional line. I'll consider it as a cleanup step.

barneygale replied:
Probably not worth it given this is the only access IMO.

picnixz replied:
I've confirmed with my benchmarks that caching the attribute doesn't help (even the `chunks.append`). It would have helped in older Python versions but for recent ones (in PGO builds), it doesn't help.

dg-pb replied:
If this is insignificant (unsurprisingly), just resolve this then.

picnixz commented on Misc/NEWS.d/next/Library/2024-07-25-18-06-51.gh-issue-122288.-_xxOR.rst:
```diff
@@ -0,0 +1,2 @@
+Improve the performances of :func:`fnmatch.translate` by a factor 1.3. Patch
```
```suggestion
Improve the performances of :func:`fnmatch.translate` by a factor 1.7. Patch
```

picnixz committed:
Update Misc/NEWS.d/next/Library/2024-07-25-18-06-51.gh-issue-122288.-_xxOR.rst

dg-pb commented on Lib/fnmatch.py:
```diff
@@ -143,50 +150,40 @@ def _translate(pat, STAR, QUESTION_MARK):
+    append, extend = buffer.append, buffer.extend
+    i = j + 1
+    for j in iter_indices:
+        # Now deal with STAR fixed STAR fixed ...
```
```suggestion
        # Now deal with STAR fixed ...
```

picnixz replied:
I'd like to keep the same comment. So that we know that after a STAR it's something that is fixed.

dg-pb replied:
Does it have to be repeated twice?

picnixz replied:
This was the historical comment and I don't mind it personally. At least, it does not leave any room for ambiguities about what comes after "fixed".

picnixz committed:
use lower-case parameter names

barneygale commented on Lib/fnmatch.py:
```diff
@@ -77,24 +77,33 @@ def translate(pat):
-            add(QUESTION_MARK)
+            # Handling '?' one at a time seems to more efficient
+            # even if there are consecutive '?' that could have
+            # been written directly.
```
This comment feels unnecessary, because this whole loop works one character at a time by default, only consuming multiple characters in a single iteration when there's a good reason (e.g. parsing a `[group]`)

picnixz replied:
I wanted to keep this comment to avoid thinking that we could have optimized using a `while` as we did for the `*`. But we can remove this comment if you think it's too much (I'll do it tomorrow).

barneygale replied:
The `*` optimisation is different, because `**` is equivalent to `*` in fnmatch, but it's not the case that `??` is equivalent to `?`.

picnixz replied:
Ah maybe I wasn't clear. What I wanted to do was:

```py
elif c == '?':
    add(question_mark)
	while i < n and pat[i] == '?':
        i += 1
		add(question_mark)
```

I think I tried this pattern in C and it was faster. But in Python, it didn't appear to (or perhaps it wasn't faster at all and I'm remembering incorrectly). 

I'll remove the comment since it appears to be confusing.

barneygale commented on Lib/fnmatch.py:
```diff
@@ -77,24 +77,33 @@ def translate(pat):
+    return _join_translated_parts(parts, indices)
 
+_re_setops_sub = re.compile(r'([&~|])').sub
+_re_escape = functools.lru_cache(maxsize=32768)(re.escape)
```
Given this is only called to escape _single characters_, the maxsize seems OTT. I suggest something in the region of 256-1024.

picnixz replied:
I think we should be fine with 512.

barneygale commented on Lib/fnmatch.py:
```diff
@@ -77,24 +77,33 @@ def translate(pat):
         c = pat[i]
         i = i+1
         if c == '*':
+            # store the position of the wildcard
```
```suggestion
            # store the position of the star wildcard
```

We don't store the positions of other wildcards

picnixz replied:
When I think about wildcards, I don't think about something else than '*' but you're right that this is important here.

barneygale commented on Lib/fnmatch.py:
```diff
@@ -77,24 +77,33 @@ def translate(pat):
+def _translate(pat, star, question_mark):
     res = []
     add = res.append
+    indices = []
```
To be clear what these indices are:

```suggestion
    star_indices = []
```

(will need other changes below)

picnixz committed:
rename variable `indices` to `star_indices`

picnixz committed:
remove ambiguous comment about '?' case

picnixz committed:
change the cache size for `re.escape`

The rationale for this change is as follows:

re.escape() is only used to cache single Unicode characters
in shell patterns; we may heuristically assume that they are
ISO-8859-1 encodable, thereby requiring a cache of size 256.
To allow non-traditional glyphs (or alphabets with a small
number of common glyphs), we double the cache size.

barneygale commented on Lib/fnmatch.py:
```diff
@@ -77,24 +77,30 @@ def translate(pat):
-    parts = _translate(pat, STAR, '.')
-    return _join_translated_parts(parts, STAR)
+    parts, indices = _translate(pat, '*', '.')
+    return _join_translated_parts(parts, indices)
```
```suggestion
    parts, star_indices = _translate(pat, '*', '.')
    return _join_translated_parts(parts, star_indices)
```

barneygale reviewed (approved):
LGTM, nice work :)

picnixz committed:
Update Lib/fnmatch.py

Co-authored-by: Barney Gale <barney.gale@gmail.com>

picnixz commented on Lib/test/test_fnmatch.py:
```diff
@@ -250,6 +250,75 @@ def test_translate(self):
+                        '.', '.', '[a-z]', '[1-9]', '*',
+                        r'\+', r'\+', r'\[', r'\[', 'a']
+        self.assertListEqual(parts, expect_parts)
+        self.assertListEqual(indices, [3, 8])
```
```suggestion
    def test_star_indices_locations(self):
        from fnmatch import _translate

        blocks = ['a^b', '***', '?', '?', '[a-z]', '[1-9]', '*', '++', '[[a']
        parts, star_indices = _translate(''.join(blocks), '*', '.')
        expect_parts = ['a', r'\^', 'b', '*',
                        '.', '.', '[a-z]', '[1-9]', '*',
                        r'\+', r'\+', r'\[', r'\[', 'a']
        self.assertListEqual(parts, expect_parts)
        self.assertListEqual(star_indices, [3, 8])
```

picnixz committed:
Update Lib/test/test_fnmatch.py

picnixz committed:
Merge branch 'main' into fnmatch-py-implementation
